vertical-pod-autoscaler/updater/eviction/pods_eviction_restriction.go

```diff
@@ -0,0 +1,233 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eviction
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	api "k8s.io/kubernetes/pkg/api"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	policyv1 "k8s.io/kubernetes/pkg/apis/policy/v1beta1"
+	kube_client "k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+)
+
+// PodsEvictionRestriction controls pods evictions. Ensures that we will not evict too many pods from one replica set
+type PodsEvictionRestriction interface {
+	// Evict checks if pod can be safely evicted
+	Evict(pod *apiv1.Pod) error
+	// CanEvict sends eviction instruction to api client.
+	// Retrurns error if pod cannot be evicted or if client returned error.
+	CanEvict(pod *apiv1.Pod) bool
+}
+
+type podsEvictionRestrictionImpl struct {
+	client         kube_client.Interface
+	podsCreators   map[string]podReplicaCreator
+	evictionBudget map[podReplicaCreator]int
+}
+
+// PodsEvictionRestrictionFactory creates PodsEvictionRestriction
+type PodsEvictionRestrictionFactory interface {
+	// NewPodsEvictionRestriction creates PodsEvictionRestriction for given set of pods.
+	NewPodsEvictionRestriction(pods []*apiv1.Pod) PodsEvictionRestriction
+}
+
+type podsEvictionRestrictionFactoryImpl struct {
+	client                    kube_client.Interface
+	minReplicas               int
+	evictionToleranceFraction float64
+}
+
+type podReplicaCreator struct {
+	Namespace string
+	Name      string
+	Kind      string
+}
+
+// CanEvict checks if pod can be safely evicted
+func (e *podsEvictionRestrictionImpl) CanEvict(pod *apiv1.Pod) bool {
+	cr, present := e.podsCreators[getPodId(pod)]
+	if present {
+		return e.evictionBudget[cr] > 0
+	}
+	return false
+}
+
+// Evict sends eviction instruction to api client. Retrurns error if pod cannot be evicted or if client returned error
+// Does not check if pod was actually evicted after eviction grace period.
+func (e *podsEvictionRestrictionImpl) Evict(podToEvict *apiv1.Pod) error {
+	cr, present := e.podsCreators[getPodId(podToEvict)]
+	if !present {
+		return fmt.Errorf("Pod not suitable for eviction %v", podToEvict.Name)
```

Comment:
Add information why. Also start errors from a lower-case letter (here and elsewhere).